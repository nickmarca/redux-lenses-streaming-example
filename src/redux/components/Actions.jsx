import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Actions as KafkaActions } from 'redux-lenses-streaming';
import { Action } from '../actions';
import SQLInput from './SQLInput';
import ListActions from './ListActions';
import Search from './Search';

export class Actions extends React.Component {
    constructor(props) {
        super(props);

        this.onSqlsChange = this.onSqlsChange.bind(this);
        this.onSubscribe = this.onSubscribe.bind(this);
        this.onUnsubscribe = this.onUnsubscribe.bind(this);
        this.onClearMessages = this.onClearMessages.bind(this);

        this.state = {
            sqls: '',
            chunk: 0,
        };
    }

    componentDidUpdate() {
        const { subscriptions, totalOfMessages } = this.props;
        const { chunk } = this.state;

        if (totalOfMessages > 15000 * chunk) {
            subscriptions.map(subscription => {
                this.onUnsubscribe(subscription);
            });
        }
    }

    onSqlsChange(event) {
        this.setState({ sqls: event.target.value });
    }

    onSubscribe() {
        const request = {
            sqls: this.state.sqls,
        };
        this.props.subscribe(request);
        const { chunk } = this.state;
        this.setState({ chunk: chunk + 1 });
    }

    onClearMessages() {
        this.props.clearMessages();
        this.setState({ chunk: 0 });
    }

    onUnsubscribe(topic) {
        const request = {
            topics: [topic],
        };
        this.props.unsubscribe(request);
    }

    render() {
        const { messages, subscriptions, connection, onSerch } = this.props;
        const { sqls } = this.state;

        const topics = subscriptions.map(subscription => (
            <button
                onClick={this.onUnsubscribe.bind(this, subscription)}
                key={subscription}
                className="button is-danger is-outlined is-small is-pulled-right"
            >
                <span>{subscription}</span>
                <span className="icon is-small">
                    <i className="fa fa-times" />
                </span>
            </button>
        ));

        return (
            <nav className="ws-subscribe panel">
                <SQLInput onChange={this.onSqlsChange} value={sqls} />

                <ListActions
                    topics={topics}
                    isClearDisabled={!connection}
                    isSubscribeDisabled={!connection || !this.state.sqls}
                    onClearMessages={this.onClearMessages}
                    onSubscribe={this.onSubscribe}
                />
                <Search
                    isSearchDisabled={!connection || !messages.length}
                    onSearch={onSerch}
                />
            </nav>
        );
    }
}

Actions.defaultProps = {};

Actions.propTypes = {
    subscribe: PropTypes.func.isRequired,
    unsubscribe: PropTypes.func.isRequired,
    onSerch: PropTypes.func.isRequired,
    totalOfMessages: PropTypes.number.isRequired,
};

const mapStateToProps = state => ({
    subscriptions: state.lenses.subscriptions,
    connection: state.lenses.connection,
    messages: state.session.messages,
});

const mapDispatchToProps = {
    ...KafkaActions,
    ...Action,
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Actions);
