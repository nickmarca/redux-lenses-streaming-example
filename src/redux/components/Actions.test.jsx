import React from 'react';
import { create } from 'react-test-renderer';
import { Actions } from './Actions';

describe('Actions', () => {
    it('matches the snapshot', () => {
        const onSearch = jest.fn();
        const subscribe = jest.fn();
        const unsubscribe = jest.fn();
        const actions = create(
            <Actions
                subscriptions={[]}
                onSerch={onSearch}
                subscribe={subscribe}
                totalOfMessages={0}
                unsubscribe={unsubscribe}
            />
        );

        expect(actions.toJSON()).toMatchSnapshot();
    });
});
