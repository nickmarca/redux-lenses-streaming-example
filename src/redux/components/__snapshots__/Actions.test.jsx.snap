// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions matches the snapshot 1`] = `
<nav
  className="ws-subscribe panel"
>
  <div
    className="panel-heading"
  >
    <div
      className="field has-addons"
    >
      <p
        className="control is-expanded"
      >
        <textarea
          className="textarea is-small is-info"
          onChange={[Function]}
          placeholder="SQLS"
          rows="3"
          value=""
        />
      </p>
    </div>
  </div>
  <div
    className="panel-block"
  >
    <div
      className="control"
    >
      <button
        className="button is-small is-info"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "10px",
          }
        }
      >
        Subscribe
      </button>
      <button
        className="button is-small is-danger"
        disabled={true}
        onClick={[Function]}
      >
        Clear Messages
      </button>
    </div>
    <div
      className="control"
    />
  </div>
  <div
    className="panel-block"
  >
    <div
      className="field has-addons"
    >
      <div
        className="control"
      >
        <input
          className="input is-small"
          onChange={[Function]}
          placeholder="Find a message"
          type="text"
          value=""
        />
      </div>
      <div
        className="control"
      >
        <button
          className="button is-info is-small"
          disabled={true}
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </div>
</nav>
`;
